ruby.asciidoc

```diff
@@ -0,0 +1,565 @@
+directory:
+
+[source,shell]
+----
+$ ruby rectangle.rb
+----
+
+And the Pony version:
+
+[source,ponylang]
+----
+use "collections"
+
+actor Main
+  new create(env: Env) =>
+    for i in Range[I32](1, 5) do
+        let r = Rectangle(i, i + 2)
+        env.out.print("Width and height: " + r.dimensions())
+        env.out.print("Circumference: " + r.circumference().string())
+        env.out.print("Area: " + r.area().string())
+        env.out.print("")
+    end
+
+class Rectangle
+  let _width: I32
+  let _height: I32
+
+  new create(width: I32, height: I32) =>
+    _width = width
+    _height = height
+
+  fun dimensions(): String => _width.string() + " " + _height.string()
+
+  fun circumference(): I32 => 2 * (_width + _height)
+
+  fun area(): I32 => _width * _height
+----
+
+In the Pony version, the `Rectangle` class structure is essentially the
+same as the Ruby version. The `let` keyword introduces fields or local
+variables that must be initialized exactly once. Visibility in Pony is
+controlled by the naming: elements with a leading underscore are package
+private, all others are public.
+
+Because Pony is also a functional language, all statements are also
+expressions. The return value of functions is simply the last expression
+in the body. This is really similar to Ruby.
+
+For-loops in Pony are a lot like Python’s for loop, and is controlled by
+an intance of the `Iterator` interface. In this case we want a simple
+integer range, so we use a `Range` object parameterized with `I32` which
+gives us a sequential range of 32-bit integers.
+
+The `Range` class is in the collections package, and we import that with
+the `use` keyword.
+
+Unlike Ruby with its automatic String conversions using `to_s`, Pony
+doesn’t have any implicit conversions: we have to invoke the `string()`
+method on each of the numeric values to get strings that we can
+concatenate to build the output.
+
+To run the program, execute these commands in the measurements
+directory:
+
+[source,shell]
+----
+$ ponyc main.pony
+$ ./measurements
+----
+
+[[getting-functional]]
+Getting functional
+~~~~~~~~~~~~~~~~~~
+
+In this example we just want to use a pure function to return the
+comparison of two integers as a human readable string.
+
+[source,ruby]
+----
+cat functional/functional.rb
+
+def compare(a, b)
+  if (a > b)
+    "a is bigger than b"
+  elsif (a < b)
+    "b is bigger than a"
+  else
+    "a and b are the same"
+  end
+end
+
+puts compare(1, 2)
+----
+
+In the Ruby version, we use a method to avoid needing an instance.
+
+Each comparison and corresponding return is computed with an imperative
+if and return
+
+To run the program, execute this command in the functional directory:
+
+[source,shell]
+----
+$ ruby functional.rb
+----
+
+[source,ponylang]
+----
+$ cat functional/main.pony
+
+actor Main
+  new create(env : Env) =>
+    env.out.print(compare(2,1))
+
+  fun compare(a: I64, b: I64): String =>
```

Comment:
If you want to demonstrate an idiomatic "functional" style in Pony, I'd recommend making this a `primitive` with an `apply` method, instead of making it a method of the `Main` actor.

```pony
primitive Compare
  fun apply(a: I64, b: I64): String =>
    // ...
```

The resulting primitive can be called like this:
```pony
env.out.print(Compare(1, 2))
```

This also gives a good opportunity to show off the `apply` syntax sugar.